impl/src/main/java/com/okta/sdk/impl/ds/cache/DefaultResourceCacheStrategy.java

```diff
@@ -0,0 +1,236 @@
+        }
+
+        // return if no data
+        if (Collections.isEmpty(data)) {
+            return null;
+        }
+
+        return new DefaultResourceDataResult(request.getAction(), uri, clazz, coerce(data));
+    }
+
+    @SuppressWarnings("unchecked")
+    private void cache(Class<? extends Resource> clazz, Map<String, ?> data, CanonicalUri uri) {
+
+        Assert.notEmpty(data, "Resource data cannot be null or empty.");
+        String href = hrefResolver.resolveHref(data, clazz, baseUrlResolver.getBaseUrl());
+
+        if (isDirectlyCacheable(clazz, data)) {
+            Assert.notNull(href, "Resource data must contain an 'href' attribute.");
+            Assert.isTrue(data.size() > 1, "Resource data must be materialized to be cached " +
+                    "(need more than just an 'href' attribute)."); // TODO: this likely is not valid for Okta
+        }
+
+        //create a map to reflect the resource's canonical representation - this is what will be cached:
+        Map<String, Object> cacheValue = cacheMapInitializer.initialize(clazz, data, uri.getQuery());
+
+        for (Map.Entry<String, ?> entry : data.entrySet()) {
+
+            String name = entry.getKey();
+            Object value = entry.getValue();
+
+            cacheValue.put(name, value);
+        }
+
+        if (isDirectlyCacheable(clazz, cacheValue)) {
+            Cache cache = getCache(clazz);
+            String cacheKey = getCacheKey(href);
+            Object previousCacheValue = cache.put(cacheKey, cacheValue);
+            logger.debug("Caching object for key '{}', class: '{}', updated {}", cacheKey, clazz, previousCacheValue != null);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void uncache(String cacheKey, Class<? extends Resource> resourceType) {
+        Assert.hasText(cacheKey, "cacheKey cannot be null or empty.");
+        Assert.notNull(resourceType, "resourceType cannot be null.");
+        Cache<String, Map<String, ?>> cache = getCache(resourceType);
+        cache.remove(cacheKey);
+        logger.debug("Removing cache for key '{}', class: '{}'", cacheKey, resourceType);
+    }
+
+    private boolean isCacheable(ResourceDataResult result) {
+        if (Collections.isEmpty(result.getData())) {
+            return false;
+        }
+
+        Class<? extends Resource> clazz = result.getResourceClass();
+
+        //@since 0.5.0
+        boolean materialized = isMaterialized(result.getData(), clazz);
+
+        if (!materialized) {
+            logger.debug("Class: {}, is not cacheable.", clazz.getSimpleName());
+        }
+
+        return materialized;
+    }
+
+    /**
+     * Quick fix for <a href="https://github.com/stormpath/stormpath-sdk-java/issues/17">Issue #17</a>.
+     *
+     */
+    private boolean isDirectlyCacheable(Class<? extends Resource> clazz, Map<String, ?> data) {
+
+        return isMaterialized(data, clazz) &&
+               // do NOT cache collections
```

Comment:
Why not?

Reply:
A lot more logic and testing is required before we can do this. SP had experimental support for this but it was never enabled.  Some resources differ between the object returned from the cache and the full object, and some of the collections change based on serve side events.

Its very possible, and it would be great for some objects (like caching the list of a user's groups), we just don't have a concrete list which ones we should support yet.